near-sdk/src/test_utils/mod.rs

```diff
@@ -7,10 +7,68 @@ mod context;
 pub use context::{accounts, testing_env_with_promise_results, VMContextBuilder};
 use near_vm_logic::mocks::mock_external::Receipt;
 
+/// Initializes a testing environment to mock interactions which would otherwise go through a
+/// validator node. This macro will initialize or overwrite the [`BLOCKCHAIN_INTERFACE`]
+/// instance which satisfies the [`BlockchainInterface`] trait for interactions from a
+/// smart contract.
+///
+/// There are five parameters that can be accepted to configure the interface with a
+/// [`MockedBlockchain`], in this order:
+/// - `context`: [`VMContext`] which contains some core information about
+/// the blockchain and message data which can be used from the smart contract.
+/// - `config` (optional): [`VMConfig`] which contains some additional information
+/// about the VM to configure parameters not directly related to the transaction being executed.
+/// - `fee_config`(optional): [`RuntimeFeesConfig`] which configures the
+/// fees for execution and storage of transactions.
+/// - `validators`(optional): a [`HashMap`]<[`AccountId`], [`Balance`]> mocking the
+/// current validators of the blockchain.
+/// - `promise_results`(optional): a [`Vec`] of [`PromiseResult`] which mocks the results
+/// of callback calls during the execution.
+///
+/// Any argument not included will use the default implementation of each.
+///
+/// # Example use
+///
+/// ```
+/// #[macro_use]
+/// extern crate near_sdk;
```

Comment:
This ideally should be `use near_sdk::testing_env`, `extern crate` shouldn't be used in modern Rust code. Though, it also requires that the macro uses `$crate` internally and doesn't just rely on everything being in scope.

Reply:
there was a reason I couldn't do that, I'll try again to see if I was wrong or to pinpoint what that issue was

Reply:
Yeah, it's true that `near_sdk` could be written in such a way that extern crate is required. I *think* it should be possible to fix that in a backwards compatible way. 

I think this'll do the trick:

```rust
diff --git a/near-sdk/src/test_utils/mod.rs b/near-sdk/src/test_utils/mod.rs
index 3506e20..a8ac175 100644
--- a/near-sdk/src/test_utils/mod.rs
+++ b/near-sdk/src/test_utils/mod.rs
@@ -68,12 +68,12 @@ use near_vm_logic::mocks::mock_external::Receipt;
 #[macro_export]
 macro_rules! testing_env {
     ($context:expr, $config:expr, $fee_config:expr, $validators:expr, $promise_results:expr $(,)?) => {
-        near_sdk::env::set_blockchain_interface(Box::new(near_sdk::MockedBlockchain::new(
+        $crate::env::set_blockchain_interface(Box::new($crate::MockedBlockchain::new(
             $context,
             $config,
             $fee_config,
             $promise_results,
-            match near_sdk::env::take_blockchain_interface() {
+            match $crate::env::take_blockchain_interface() {
                 Some(mut bi) => bi.as_mut_mocked_blockchain().unwrap().take_storage(),
                 None => Default::default(),
             },
@@ -82,14 +82,14 @@ macro_rules! testing_env {
         )));
     };
     ($context:expr, $config:expr, $fee_config:expr, $validators:expr $(,)?) => {
-        testing_env!($context, $config, $fee_config, $validators, Default::default());
+        $crate::testing_env!($context, $config, $fee_config, $validators, Default::default());
     };
 
     ($context:expr, $config:expr, $fee_config:expr $(,)?) => {
-        testing_env!($context, $config, $fee_config, Default::default());
+        $crate::testing_env!($context, $config, $fee_config, Default::default());
     };
     ($context:expr) => {
-        testing_env!($context, Default::default(), Default::default());
+        $crate::testing_env!($context, Default::default(), Default::default());
     };
 }
 
```

Reply:
hmm, yeah I don't see an issue with this. I remember I tried this, but I might have scrapped because I wasn't sure on the implications and didn't want to have any breaking changes. I'll commit this change now